### ISSUE TITLE
Dev dependency is installed with --no-dev option 

### ISSUE THREAD

**Issue by calvinalkan** [composer/composer#10913]
I believe I have found a bug where Composer silently downloads a transitive dev dependency even tho --no-dev is specified.
This happens when one of the required packages declares a dependency on a psr*-implementation and you did not provide one + one of your dev dependencies has a transitive dependency on that psr*-implemenation. 

The correct behavior here should be to abort the installation.

I created a reproducer repo (https://github.com/calvinalkan/composer-bug). This took me hours to find out why I had packages in my build that I did not declare anywhere. Verified with the latest composer version (2.3.8) and the official composer docker image.

( I cant adhere to the commit template because GitHub is giving me "Comment is too long (maximum is 65536 characters)")
All the output and steps to reproduce are in the reproducer repo.

**Comment by GuySartorelli** [comment]
What is the output of `composer why guzzlehttp/psr7`?

**Comment by calvinalkan** [comment]
ˋ There is no installed package depending on "guzzlehttp/psr7"ˋ

please have a look at the reproducer repo I created. I documented all steps to reproduce this.

https://github.com/calvinalkan/composer-bug

**Comment by Seldaek** [comment]
This explains it:

```
$ c why psr/http-message-implementation -t
guzzlehttp/psr7 2.4.0 PSR-7 message implementation that also provides common utility methods
└──laminas/laminas-httphandlerrunner 2.1.0 (requires psr/http-message-implementation ^1.0)
   └──composer/bug dev-master (requires laminas/laminas-httphandlerrunner 2.1.0)
```

The laminas package requires psr/http-message-implementation - which is provided by guzzlehttp/psr7, so that one package is needed even in non-dev where the rest of guzzle is not needed.

It'd be nice if `composer why` would show this so I'll leave this open.

**Comment by stof** [comment]
@calvinalkan no. `composer install` will not fail. `composer update` will fail when updating the dependencies, asking you to provide a `psr/http-message-implementation`. Composer will not let you generate a lock file without that implementation that would then break `composer install` (at least, Composer 2 will not, but we don't care about Composer 1 issues anymore).
